js/src/ValidateSPV.js

```diff
@@ -250,3 +250,126 @@ export function validateHeaderChain(headers) {
+ * @param {Uint8Array}    hashLE The LE hash of the header
+ * @param {Number}        height The height
+ * @param {Uint8Array}    merkleRoot The merkle root of the header
+ * @param {Uint8Array}    merkleRootLE The LE merkle root
+ * @param {Uint8Array}    prevHash The hash of the previous header
+ * @returns {Boolean}     True if all the elements of the header are valid
+ * @throws {Error}        If any of the bitcoin header elements are invalid
+*/
+export function validateHeader(
+  header,
+  hash,
+  hashLE,
+  height,
+  merkleRoot,
+  merkleRootLE,
+  prevHash
+) {
+  // Check that HashLE is the correct hash of the raw header
+  const headerHash = BTCUtils.hash256(header);
+  if (!utils.typedArraysAreEqual(headerHash, hashLE)) {
+    throw new Error('Hash LE is not the correct hash of the header');
+  }
+
+  // Check that HashLE is the reverse of Hash
+  const reversedHash = utils.reverseEndianness(hash);
+  if (!utils.typedArraysAreEqual(reversedHash, hashLE)) {
+    throw new Error('HashLE is not the LE version of Hash');
+  }
+
+  // Check that the MerkleRootLE is the correct MerkleRoot for the header
+  const extractedMerkleRootLE = BTCUtils.extractMerkleRootLE(header);
+  if (!utils.typedArraysAreEqual(extractedMerkleRootLE, merkleRootLE)) {
+    throw new Error('MerkleRootLE is not the correct merkle root of the header');
+  }
+
+  // Check that MerkleRootLE is the reverse of MerkleRoot
+  const reversedMerkleRoot = utils.reverseEndianness(merkleRoot);
+  if (!utils.typedArraysAreEqual(reversedMerkleRoot, merkleRootLE)) {
+    throw new Error('MerkleRootLE is not the LE version of MerkleRoot');
+  }
+
+  // Check that PrevHash is the correct PrevHash for the header
+  const extractedPrevHash = BTCUtils.extractPrevBlockBE(header);
+  if (!utils.typedArraysAreEqual(extractedPrevHash, prevHash)) {
+    throw new Error('Prev hash is not the correct previous hash of the header');
+  }
+
+  return true;
+}
+
+/**
+ *
+ * Checks validity of an entire SPV Proof
+ *
+ * @dev                   Checks that each element in an SPV Proof is valid
+ * @param {Uint8Array}    version The version
+ * @param {Uint8Array}    vin The vin
+ * @param {Uint8Array}    vout The vout
+ * @param {Uint8Array}    locktime The locktime
+ * @param {Uint8Array}    txid The tx ID
+ * @param {Uint8Array}    txidLE The LE tx ID
+ * @param {Number}        index The index
+ * @param {Uint8Array}    intermediateNodes The intermediate nodes
+ * @param {Uint8Array}    header The bitcoin header
+ * @param {Uint8Array}    hash The hash of the header
+ * @param {Uint8Array}    hashLE The LE hash of the header
+ * @param {Number}        height The height
+ * @param {Uint8Array}    merkleRoot The merkle root of the header
+ * @param {Uint8Array}    merkleRootLE The LE merkle root
+ * @param {Uint8Array}    prevHash The hash of the previous header
+ * @returns {Boolean}     True if all the elements of the proof are valid, false if otherwise
+*/
+export function validateProof(
```

Comment:
can we make this accept the deserialized Proof object rather than all of its attributes?

Reply:
Done